File: telegram-service/src/main/java/de/nak/home_assistant/commands/Fritz.java
package de.nak.home_assistant.commands;

import com.pengrad.telegrambot.TelegramBot;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.request.SendMessage;
import de.nak.home_assistant.CustomKeyboard;
import de.nak.home_assistant.actions.DeviceList;
import de.nak.home_assistant.models.database.ModuleResponse;
import de.nak.home_assistant.models.telegram.Command;
import de.nak.home_assistant.services.database.ModuleService;
import de.nak.home_assistant.services.database.UserService;

import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Optional;

import static de.nak.home_assistant.Utils.messageArgs;

public class Fritz extends Command {

    private final TelegramBot bot;

    public Fritz(TelegramBot bot) {
        this.bot = bot;
        this.name = "fritz";
        this.description = "Öffnet das fritz menü";
    }

    public Fritz() {
        this(null);
    }

    public void reply(Update update) {
        long userId = update.message().from().id();
        long chatId = update.message().chat().id();

        Optional<ModuleResponse> moduleResponse = new ModuleService().findModuleByName(this.name);
        if (!moduleResponse.isPresent()) {
            SendMessage msg = new SendMessage(chatId, "Interner Fehler (es wurde kein Modul mit dem Namen " + this.name + " gefunden)");
            bot.execute(msg);
            return;
        }

        int moduleId = moduleResponse.get().getId();
        List<SendMessage> messages = new ArrayList<>();

        String[] args = messageArgs(update.message().text(), this.name);
        boolean hasArgs = args.length > 0;
        boolean tokenSuccess = false;
        UserService userService = new UserService(userId);


        if (hasArgs) {
            byte[] token = args[0].getBytes();
            tokenSuccess = userService.setToken(moduleId, Base64.getEncoder().encodeToString(token));
        }

        // Send success message
        if (tokenSuccess) {
            messages.add(new SendMessage(chatId, "Fritz-Token wurde geupdated 🥳"));
        }

        // Remove module and send error if token couldn't be updated
        if (!tokenSuccess && hasArgs) {
            SendMessage msg = CustomKeyboard.generateMessageWithKeyboard(userId, chatId,"Fehler beim Token update :(");
            messages.add(msg);
        }

        // Generate Message if token was updated or shouldn't be updated
        if (tokenSuccess || !hasArgs) {
            messages.add(DeviceList.generateMessage(userId, chatId, moduleId));
        }

        messages.forEach(bot::execute);
    }
}

File: packages/telegram-service/src/commands/fritz.ts
import { Message } from "node-telegram-bot-api";
import { sendDeviceList } from "../actions/device-list";
import { bot, Command } from "../bot";
import { i18n } from "../i18n";
import { findModuleByName } from "../services/module";
import { setToken } from "../services/user";

export const Fritz: Command = () => {
  const { fritz } = i18n();

  return {
    command: fritz.name,
    regex: new RegExp(`/?${fritz.name} ?(.+)?`),
    description: fritz.description,
    handler: async (msg, match) => {
      if (!match) {
        return;
      }

      return replyToFritz(msg, match);
    },
  };
};

export const replyToFritz = async (
  { from, chat }: Message,
  match: RegExpMatchArray
): Promise<void> => {
  if (!from) {
    return;
  }

  const translations = i18n();

  const module = await findModuleByName(Fritz.name);

  if (module) {
    const hasArgs = Boolean(match[1]);
    let tokenSuccess = false;

    if (hasArgs) {
      await setToken(
        from.id,
        module.id,
        Buffer.from(match[1]).toString("base64")
      )
        .then(() => {
          tokenSuccess = true;
          return bot.sendMessage(chat.id, translations.fritz.tokenUpdate);
        })
        .catch(() => bot.sendMessage(chat.id, translations.fritz.tokenError));
    }

    if (tokenSuccess || !hasArgs) {
      await sendDeviceList(from.id, chat.id, module.id);
      return;
    }
  }

  await bot.sendMessage(chat.id, translations.fritz.moduleError);
};
